samples/algorithms/sudoku-grover/ColoringGroverWithConstraints.qs

```diff
@@ -0,0 +1,395 @@
+    ///    |   | 6 | 2 | 7 | 8 | 3 | 4 | 0 | 1 |
+    ///    -------------------------------------
+    ///    | 8 |   | 1 | 6 | 2 | 4 | 3 | 7 | 5 |
+    ///    -------------------------------------
+    ///    | 7 | 3 | 4 | 5 | 0 | 1 | 8 | 6 | 2 |
+    ///    -------------------------------------
+    ///    | 6 | 8 | 7 | 1 | 5 | 0 | 2 | 4 | 3 |
+    ///    -------------------------------------
+    ///    | 4 | 1 | 5 | 3 | 6 | 2 | 7 | 8 | 0 |
+    ///    -------------------------------------
+    ///    | 0 | 2 | 3 | 4 | 7 | 8 | 1 | 5 | 6 |
+    ///    -------------------------------------
+    ///    | 3 | 5 | 8 | 0 | 1 | 7 | 6 | 2 | 4 |
+    ///    -------------------------------------
+    ///    | 1 | 7 | 6 | 2 | 4 | 5 | 0 | 3 | 8 |
+    ///    -------------------------------------
+    ///    | 2 | 4 | 0 | 8 | 3 | 6 | 5 | 1 | 7 |
+    ///    -------------------------------------
+    ///  The challenge is to fill the empty squares with numbers 0 to 8
+    ///  that are unique in row, column and the top left 3x3 square
+    ///  This is a graph coloring problem where the colors are 0 to 8
+    ///  and the empty cells are the vertices. The vertices can be defined as   
+    ///     -----------------
+    ///     | 0 |   |   |   | ...
+    ///     -----------------
+    ///     |   | 1 |   |   | ...
+    ///     -----------------
+    ///     |   |   |   |   | ...
+    ///     ...
+    /// The graph is
+    ///  0---1
+    /// But we also require that 
+    ///    - vertex 0 can not have value 6,2,7,8,3,4,0,1 (row constraint)
+    ///                         or value 8,7,6,4,0,3,1,2 (col constraint)
+    ///    - vertex 1 can not value 8,1,6,2,4,3,7,5 (row constraint)
+    ///                    or value 6,3,8,1,2,5,7,4 (col constraint)
+    /// This results in edges (vertices that can not be same color)
+    /// edges = [(1, 0)]
+    /// This is saying that vertex 1 can not have same color as vertex 0
+    /// and startingColorConstraints = [(0, 8),(0, 7),(0, 6),(0, 4),(0, 0),(0, 3),
+    ///     (0, 1),(0, 2),(1, 6),(1, 3),(1, 8),(1, 1),(1, 2),(1, 5),(1, 7),(1, 4)]
+    /// The colors found must be from 0 to 8, which requires 4 bits per color
+    /// A valid graph coloring solution is: [5,0]
+    /// i.e. vextex 0 has color 5, vertex 1 has color 0
+    operation VertexColoringOracle4Bit9Color (numVertices : Int, edges : (Int, Int)[],  
+        startingColorConstraints : (Int, Int)[], 
+        colorsRegister : Qubit[], target : Qubit) : Unit is Adj+Ctl {
+        let nEdges = Length(edges);
+        let bitsPerColor = 4; // 4 bits per color
+        let nStartingColorConstraints = Length(startingColorConstraints);
+        // we are looking for a solution that 
+        // (a) has no edge with same color at both ends and 
+        // (b) has no Vertex with a color that violates the starting color constraints
+        using (conflictQubits = Qubit[nEdges + nStartingColorConstraints + numVertices]) {
```

Comment:
Similarly here, defining multiple registers will eliminate the need for a lot of index gymnastics below.